src/main/java/io/BinaryCurveReader.java

```diff
@@ -0,0 +1,238 @@
+package io;
+
+import algebra.curves.AbstractG1;
+import algebra.curves.AbstractG2;
+import algebra.fields.AbstractFieldElement;
+import algebra.fields.AbstractFieldElementExpanded;
+import common.PairRDDAggregator;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.function.Supplier;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import scala.Tuple2;
+
+/**
+ * Base class for binary readers. Handles reading fixed-width BigInteger values (in big-endian
+ * form).
+ */
+public abstract class BinaryCurveReader<
+        FrT extends AbstractFieldElementExpanded<FrT>,
+        G1T extends AbstractG1<G1T>,
+        G2T extends AbstractG2<G2T>>
+    extends DataInputStream implements AbstractCurveReader<FrT, G1T, G2T> {
+
+  protected BinaryCurveReader(InputStream inStream_) {
+    super(inStream_);
+  }
+
+  /**
+   * Version of readFr which returns null instead of throwing on error (for use in functional-style
+   * function composition.
+   */
+  public FrT readFrNoThrow() {
+    try {
+      return readFr();
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  /**
+   * Version of readG1 which returns null instead of throwing on error (for use in functional-style
+   * function composition.
+   */
+  public G1T readG1NoThrow() {
+    try {
+      return readG1();
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  /**
+   * Version of readG2 which returns null instead of throwing on error (for use in functional-style
+   * function composition.
+   */
+  public G2T readG2NoThrow() {
+    try {
+      return readG2();
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  protected BigInteger readBigInteger(final int numBytes) throws IOException {
+    final byte[] bytes = readNBytes(numBytes);
```

Comment:
Shouldn't `readNBytes` be the method of an `InputStream`? On which stream are you reading here? (why don't you need to do `inStream_.readNBytes()` here?)

Reply:
BinaryCurveReader extends the DataStream (which supports reading ints, longs, etc), so readNBytes can be called on this class.

Reply:
I see, makes sense, thanks!